site/blog/updates/why-sql-for-infrastructure.mdx

```diff
@@ -0,0 +1,482 @@
+
+-- Associate the instance with the security group
+INSERT INTO instance_security_groups (instance_id, security_group_id) SELECT
+  (SELECT id FROM instance WHERE tags ->> 'name' = 'login'),
+  (SELECT id FROM security_group WHERE group_name='login-sg' AND region = 'us-west-2');
+```
+
+We can then `SELECT * FROM iasql_preview();` to see what this does, like a `terraform plan`:
+
+:::note `iasql_preview` output
+| action | table_name | id | description |
+| :----- | :--------- | -: | ----------: |
+| create | instance | 1 | 1 |
+| create | security_group | 18 | 18 |
+| create | security_group_rule | 1 | 1 |
+:::
+
+This is a "high level" overview with minimal details. We can get back a listing of all changes, represented in auto-generated SQL with `SELECT * FROM iasql_get_sql_for_transaction();`
+
+```sql
+INSERT INTO security_group (description, group_name, region)
+VALUES ('Login security group', 'login-sg', (SELECT region FROM aws_regions WHERE region = 'us-west-2'));
+
+
+INSERT INTO security_group_rule (is_egress, ip_protocol, from_port, to_port, cidr_ipv4, region, security_group_id)
+VALUES ('f', 'tcp', '22', '22', '0.0.0.0/0', (SELECT region FROM aws_regions WHERE region = 'us-west-2'), (SELECT id FROM security_group WHERE group_id = NULL AND region = (SELECT region FROM aws_regions WHERE region = 'us-west-2')));
+
+
+INSERT INTO instance (ami, instance_type, key_pair_name, state, tags, hibernation_enabled, region, subnet_id)
+VALUES ('resolve:ssm:/aws/service/canonical/ubuntu/server/20.04/stable/current/amd64/hvm/ebs-gp2/ami-id', 't3.small', 'login', 'running', '{"name":"login"}'::jsonb, 'f', (SELECT region FROM aws_regions WHERE region = 'us-west-2'), (SELECT id FROM subnet WHERE subnet_id = 'subnet-06140fd708a495450' AND region = (SELECT region FROM aws_regions WHERE region = 'us-west-2')));
+
+
+INSERT INTO instance_security_groups (instance_id, security_group_id)
+VALUES ((SELECT id FROM instance WHERE instance_id = NULL AND region = (SELECT region FROM aws_regions WHERE region = 'us-west-2')), (SELECT id FROM security_group WHERE group_id = NULL AND region = (SELECT region FROM aws_regions WHERE region = 'us-west-2')));
+```
+
+which in this case is very similar to (though a bit more verbose than) our original SQL statements, but if you were going back and forth on what exact changes you're going to make, it can be a good summary.
+
+If we're good with this, we can `SELECT iasql_commit()` to push this to our cloud account and then update our `ssh_credentials` record with the public IP address and check the connection:
+
+```sql
+WITH h AS (
+  SELECT host(im.public_ip_address) AS hostname
+  FROM instance_metadata im
+  INNER JOIN instance i ON im.instance_id = i.instance_id
+  WHERE i.tags ->> 'name' = 'login'
+  LIMIT 1
+)
+UPDATE ssh_credentials SET hostname = h.hostname FROM ssh_credentials, h WHERE hostname = 'tbd';
+
+SELECT * FROM ssh_exec('login', 'uname -a');
+```
+
+## SQL is the best option for Infrastructure Management
```

Comment:
```suggestion
## SQL as the best of the worst option for Infrastructure Management
```

a bit nihilistic or cynical, but keeps the tone from the beginning?

Reply:
"the best of the worst" reads a bit weird to me. I think this would work a bit better?

```suggestion
## SQL is the least bad option for Infrastructure Management
```

Reply:
that is better!